packages/client/lib/sync/fetcher/blockfetcher.ts

```diff
@@ -1,60 +1,45 @@
+const level = require('level')
 import { Fetcher, FetcherOptions } from './fetcher'
 import { Block, BlockBodyBuffer } from '@ethereumjs/block'
-import { BN } from 'ethereumjs-util'
 import { Peer } from '../../net/peer'
 import { EthProtocolMethods } from '../../net/protocol'
-import { Chain } from '../../blockchain'
-
-export interface BlockFetcherOptions extends FetcherOptions {
-  /* Blockchain */
-  chain: Chain
-
-  /* Block number to start fetching from */
-  first: BN
-
-  /* How many blocks to fetch */
-  count: BN
-}
+import VM from '@ethereumjs/vm'
+import { DefaultStateManager } from '@ethereumjs/vm/dist/state'
+import { SecureTrie as Trie } from '@ethereumjs/trie'
 
 /**
  * Implements an eth/62 based block fetcher
  * @memberof module:sync/fetcher
  */
 export class BlockFetcher extends Fetcher {
-  protected chain: Chain
-  protected first: BN
-  protected count: BN
+  public vm: VM
 
   /**
    * Create new block fetcher
-   * @param {BlockFetcherOptions}
+   * @param {FetcherOptions}
    */
-  constructor(options: BlockFetcherOptions) {
+  constructor(options: FetcherOptions) {
     super(options)
 
-    this.chain = options.chain
-    this.maxPerRequest = options.maxPerRequest ?? 128
-    this.first = options.first
-    this.count = options.count
-  }
+    if (!this.config.vm) {
+      const db = level('./statedir')
+      const trie = new Trie(db)
 
-  /**
-   * Generate list of tasks to fetch
-   * @return {Object[]} tasks
-   */
-  tasks(): object[] {
-    const { first, count } = this
-    const max = this.maxPerRequest
-    const tasks = []
-    while (count.gten(max)) {
-      tasks.push({ first: first.clone(), count: max })
-      first.iaddn(max)
-      count.isubn(max)
-    }
-    if (count.gtn(0)) {
-      tasks.push({ first: first.clone(), count: count.toNumber() })
+      const stateManager = new DefaultStateManager({
+        common: this.config.common,
+        trie,
+      })
+
+      this.vm = new VM({
```

Comment:
I don't think I like that if we use `BlockFetcher`, we also require that we start running blocks. This would probably lead to problems when we implement Beam Sync. I think it makes more sense in `FullSync`.

What we could do to ensure we only run one block at a time, is to add an option in `VM.runBlockchain` which describes how many blocks we want to run (default: all), which sets `maxBlocks` of `blockchain.iterator`. When we putBlocks, in the `Chain` of `Client`, we can attach to the `updated` event, and then start running these new blocks one-by-one.

Reply:
The reason why I did this is that I wanted to have this as close as possible to the `blockchain.putBlock()` call (I even thought that the VM execution might actually be suited best to be placed in the `Blockchain` class itself (yes - I know - that atm this would cause a circular dependency). I think to have this connected by an event (`updated`) is not enough, this needs be connected in both directions. So if a VM execution fails, this should also stop block processing, otherwise it is a large waste of resources if blocks are continuedly put in the blockchain when some prior block failed in the VM.

I haven't thought about the beam sync implications though.

Would it be a way to leave this like this for now and have this open for refactoring later? I am just happy that this is working right now and my fear is that we bring this back in a non-working state or at least introduce new problems if we move here (you can proof me wrong though with a PR here).

Reply:
(so please always make sure by mainnet test runs that both initial sync from 0 and continued sync from some blocks already stored is still working including VM execution 😁 )

Reply:
Well one thing we'd need is a lot of extra tests. Since we don't ban a peer currently if they feed us wrong blocks, this implies that the current (bad) peer is still the "best peer" and will thus be used as the (only) peer to sync these (bad) blocks. So I think it would keep running these bad blocks. It is a good point though, have not thought about the mechanism on what to do if peers feed us bad blocks. You are right that there needs to be a two-way-direction here, we want VM to report bad peers, but then we'd indeed need to know which peer reported that block.